### PR TITLE
Add a date range field for project start/end dates

### PR DESCRIPTION
**Pull request by mattandrews**
This isn't fully complete yet (needs a bit of attention from @LynseyJReynolds to work out how best to display the two fields, and perhaps add an option to mark them both as the same for single-date projects), but the approach is ready to be reviewed.

image

image

image

image

image

### CODE REVIEW

File: controllers/apply-next/lib/formatters.js [hunk 1]
```diff
@@ -70,6 +70,19 @@ function formatDate(value) {
...
 
+function formatDateRange(value) {
+    const dates = {
+        start: fromDateParts(value.start),
+        end: fromDateParts(value.end)
+    };
+    if (!dates.start.isValid() && !dates.end.isValid()) {
+        return '';
+    }
+    return `${dates.start.format('D MMMM, YYYY')} – ${dates.end.format(
```
Comment (davidrapson) [inline controllers/apply-next/lib/formatters.js]: 😍  Fantastic. :neckbeard: I'd typically set the en-dash in date ranges with no space not sure if that's typical or not (yes I have opinions on the use of dashes but I'll leave typos littered across the code-base, thanks for fixing those too!)

File: controllers/apply-next/joi-extensions/date-range.js [hunk 2]
```diff
@@ -0,0 +1,157 @@
+'use strict';
+const { fromDateParts } = require('../../../modules/dates');
+
+module.exports = function dateParts(joi) {
+    return {
+        name: 'dateRange',
+        base: joi.object({
+            start: {
+                day: joi
```
Comment (davidrapson) [inline controllers/apply-next/joi-extensions/date-range.js]: You could probably make this a variable, e.g.

```js
const dateParts = {
    day: joi
        .number()
        .integer()
        .required(),
    month: joi
        .number()
        .integer()
        .required(),
    year: joi
        .number()
        .integer()
        .required()
};
...
joi.object({  start: dateParts, end: dateParts })
```

File: controllers/apply-next/lib/validate-model.js [hunk 3]
```diff
@@ -14,6 +14,7 @@ module.exports = function validateModel(formModel) {
                 'checkbox',
                 'currency',
                 'date',
+                'dateRange',
```
Comment (davidrapson) [inline controllers/apply-next/lib/validate-model.js]: Could we either make this hyphenated or make the others camelCase, this is making my eye twitch 🙂

File: controllers/apply-next/simple/fields.js [hunk 4]
```diff
@@ -391,7 +391,7 @@ module.exports = function fieldsFor({ locale, data = {} }) {
                 value: ORGANISATION_TYPES.STATUTORY_BODY,
                 label: localise({ en: 'Statutory body', cy: '' }),
                 explanation: localise({
-                    en: `<p>My organsation is a public body, such as a local authority, parsih council, or police or health authority</p>`,
+                    en: `<p>My organisation is a public body, such as a local authority, parish council, or police or health authority</p>`,
```
Comment (davidrapson) [inline controllers/apply-next/simple/fields.js]: 🙈

File: controllers/apply-next/simple/fields.js [hunk 5]
```diff
@@ -622,23 +622,39 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
             get explanation() {
                 return localise({
-                    en: `<p>This date needs to be at least 12 weeks from when you plan to submit your application. If your project is a one-off event, please tell us the date of the event.</p>
-                <p><strong>For example: ${
-                    this.settings.minDateExample
-                }</strong></p>`,
+                    en: `<p>The start date needs to be at least 12 weeks from when you plan to submit your application. If your project is a one-off event, please tell us the date of the event.</p>
+                    <p><strong>For example: ${
+                        this.settings.minDateExample
+                    }</strong>. The project end date must be within ${
```
Comment (davidrapson) [inline controllers/apply-next/simple/fields.js]: I'd either write this in words (like the "12 weeks above") or format this using moment earlier on (like in your TODO comment)

File: views/components/form-fields-next/macros.njk [hunk 6]
```diff
@@ -50,52 +50,78 @@
...
+        <legend class="ff-label ff-date__legend">
+            {{ field.label }}
+        </legend>
+        {{ explanationFor(field) }}
+        {{ dayMonthYearInputs(field, includeYear = includeYear) }}
+    </fieldset>
+{% endmacro %}
+
+{% macro inputDateRange(field) %}
+    <fieldset class="ff-date ff-date--range" id="field-{{ field.name }}">
```
Comment (davidrapson) [inline views/components/form-fields-next/macros.njk]: This needs a bit of thought, not 100% sure about the grouping and semantically the "start" and "end" labels will need something to connect that text with the field (should possibly be two fieldsets). I know this is still a work in progress.

File: controllers/apply-next/joi-extensions/date-range.js (hunk 2 above)
Comment (mattandrews) [reply controllers/apply-next/joi-extensions/date-range.js]: 😊 good call!

File: controllers/apply-next/lib/validate-model.js (hunk 3 above)
Comment (mattandrews) [reply controllers/apply-next/lib/validate-model.js]: Oops yeah good call. Will hyphenate it! Got confused as the Joi object is camel cased.

File: views/components/form-fields-next/macros.njk (hunk 6 above)
Comment (mattandrews) [reply views/components/form-fields-next/macros.njk]: Good shout. Made each bit a `<fieldset>` (eg. the overall legend/explanation, the start and the end fields):
image

File: controllers/apply-next/lib/formatters.js (hunk 1 above)
Comment (mattandrews) [reply controllers/apply-next/lib/formatters.js]: Done!
image

File: controllers/apply-next/simple/fields.js (hunk 5 above)
Comment (mattandrews) [reply controllers/apply-next/simple/fields.js]: Done, ended up making the "12 weeks" bit dynamic too so we don't inadvertently break that down the line.

File: controllers/apply-next/simple/fields.js [hunk 7]
```diff
@@ -616,35 +616,46 @@ module.exports = function fieldsFor({ locale, data = {} }) {
...
                         .format('D MMMM YYYY'),
-                    minYear: dt.format('YYYY'),
-                    // @TODO should there be a human readable version of this
-                    // (eg. "one year") for labels/explanations?
+                    minYear: minStartDate.format('YYYY'),
                     maxDurationFromStart: {
                         amount: 1,
-                        units: 'years'
+                        units: 'years',
+                        label: 'one year'
```
Comment (davidrapson) [inline controllers/apply-next/simple/fields.js]: Guessing this should be translated?

File: views/components/form-fields-next/macros.njk [hunk 8]
```diff
@@ -113,14 +113,22 @@
 {% endmacro %}
 
 {% macro inputDateRange(field) %}
-    <fieldset class="ff-date ff-date--range" id="field-{{ field.name }}">
+    <fieldset class="ff-date ff-date--range" id="field-{{ field.name }}-legend">
```
Comment (davidrapson) [inline views/components/form-fields-next/macros.njk]: Not quite sure about this markup, a fieldset with only a legend in it is a bit wonky. You can _nest_ fieldsets though. Not 100% what the right markup is here, maybe this top one should be a heading and then the two fields are fieldsets.

Comment (mattandrews) [reply views/components/form-fields-next/macros.njk]: Tried it as a div/header combo, have a look now!

File: controllers/apply-next/simple/fields.js (hunk 7 above)
Comment (mattandrews) [reply controllers/apply-next/simple/fields.js]: 👍

File: controllers/apply-next/simple/fields.test.js [hunk 9]
```diff
@@ -248,7 +248,7 @@ describe('fields', () => {
         });
     });
 
-    function assertRequiredForOrganistionTypes(field, requiredTypes) {
+    function assertRequiredForOrganisationTypes(field, requiredTypes) {
```
Comment (davidrapson) [inline controllers/apply-next/simple/fields.test.js]: Thanks for fixing this but FYI I've just removed this method as it's not used.

(Hopefully my move to Webstorm will remove all future typos…)

File: views/components/form-fields-next/macros.njk (hunk 8 above)
Comment (davidrapson) [reply views/components/form-fields-next/macros.njk]: 👍 Makes sense to me

**Pull request merged**